### ISSUE TITLE
Failed to build with c++ std 17

### ISSUE THREAD

**Issue by Atliac** [pytorch/pytorch#53409]
## To Reproduce
Create a simple PyTorch project like this:
```
cmake_minimum_required(VERSION 3.0 FATAL_ERROR)
project(dcgan)

find_package(Torch REQUIRED)

add_executable(dcgan dcgan.cpp)
target_link_libraries(dcgan "${TORCH_LIBRARIES}")
set_property(TARGET dcgan PROPERTY CXX_STANDARD 14)
```
some c++ codes:
```
#include <torch/torch.h>
#include <iostream>

int main() {
  torch::Tensor tensor = torch::eye(3);
  std::cout << tensor << std::endl;
}
```
Build it. It works, of course.
Then, modify the CMakeLists.txt:
```
set_property(TARGET dcgan PROPERTY CXX_STANDARD 17)
```
Build again. There comes the failure.
I've got some errors like: _libtorch-win-shared-with-deps-debug-1.8.0+cu111\libtorch\include\c10\util\C++17.h(303): error C2872: 'std': ambiguous symbol_ . 

## Environment
 - PyTorch Version (e.g., 1.0): 1.8.0
 - OS (e.g., Linux): windows 10
 - IDE Microsoft Visual C++ 2019
 - How you installed PyTorch (`conda`, `pip`, source):LibTorch with CUDA11.1

## Additional context

cc @malfet @seemethere @walterddr

**Comment by skyline75489** [comment]
Can you provide more info on this? Such as your environment, the build log. We can't really help with a blank "failed with c++17" issue like this.

**Comment by Surybai** [comment]
> Can you provide more info on this? Such as your environment, the build log. We can't really help with a blank "failed with c++17" issue like this.

Sorry, and yes.
I've updated the info.

**Comment by skyline75489** [comment]
Thanks for the report. I can confirm this issue does exists. However it does not seem to be related to the `C++17.h` specifically. A fix same as the one in #17863 would work.

**Comment by Atliac** [comment]
I've figured it out. There is no wrong but a missing compiler option, /permissive-.
The option, /permissive-, specifies standards conformance mode to the compiler. By default, the /permissive- option is set in new projects created by Visual Studio 2017 version 15.5 and later versions, but not for Cmake projects.
So, config CMakeLists.txt with this:
```
set(CMAKE_CXX_STANDARD 17)
set(CMAKE_CXX_STANDARD_REQUIRED TRUE)
if (MSVC)
    add_compile_options(/permissive-) 
endif()
```
Or one can use another compiler, such as clang, in VS Cmake project.
